pkg/controller/datavolume-controller.go

```diff
@@ -595,24 +546,201 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
-			}
-			if !done {
-				return reconcile.Result{},
-					r.updateSmartCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, nil)
-			}
+			return r.expandTmpPvcAfterClone(log, datavolume, pvcSpec, tmpPVC, CsiClone)
+		}
+	}
+
+	if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
+		return reconcile.Result{}, nil
+	}
+
+	sourcePvcNs := datavolume.Spec.Source.PVC.Namespace
+	if sourcePvcNs == "" {
+		sourcePvcNs = datavolume.Namespace
+	}
+	r.log.V(3).Info("CSI-Clone is available")
+
+	// Get source pvc
+	sourcePvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: sourcePvcNs, Name: datavolume.Spec.Source.PVC.Name}, sourcePvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.V(3).Info("Source PVC no longer exists")
+			return reconcile.Result{}, err // r.updateCSICloneDataVolumeStatus(cdiv1.CloneSourcePVCLost, datavolume)
+		}
+		return reconcile.Result{}, err
+	}
+
+	inUse, err := r.sourceInUse(datavolume, CSICloneSourceInUse)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	populated, err := r.isSourcePVCPopulated(datavolume)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if inUse || !populated {
+		return reconcile.Result{Requeue: true},
+			r.updateCloneStatusPhase(cdiv1.CloneScheduled, datavolume, nil, CsiClone)
+	}
+
+	// here use Temp PVC!!!
+	log.Info("Creating PVC for datavolume")
+	cloneTargetPvc, err := r.newVolumeClonePVC(datavolume, sourcePvc, pvcSpec, pvcName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: cloneTargetPvc.Namespace, Name: cloneTargetPvc.Name}, pvc); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		if err := r.client.Create(context.TODO(), cloneTargetPvc); err != nil && !k8serrors.IsAlreadyExists(err) {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// PVC already exists, check for name clash
+		pvcControllerRef := metav1.GetControllerOf(cloneTargetPvc)
+		pvcClashControllerRef := metav1.GetControllerOf(pvc)
+
+		if pvc.Name == cloneTargetPvc.Name &&
+			pvc.Namespace == cloneTargetPvc.Namespace &&
+			!reflect.DeepEqual(pvcControllerRef, pvcClashControllerRef) {
+			return reconcile.Result{}, errors.Errorf("Target Pvc Name in use") //r.updateCSICloneDataVolumeStatus(cdiv1.ClonePVCNameInUse, datavolume)
+		}
 
-			// trigger transfer and next reconcile should have pvcExists == true
-			tmpPVC.Annotations[annReadyForTransfer] = "true"
-			tmpPVC.Annotations[AnnPopulatedFor] = datavolume.Name
-			if err := r.client.Update(context.TODO(), tmpPVC); err != nil {
+		if pvc.Status.Phase == corev1.ClaimBound {
+			if err := r.setCloneOfOnPvc(pvc); err != nil {
 				return reconcile.Result{}, err
 			}
+		}
+	}
+
+	return reconcile.Result{Requeue: true}, r.updateCloneStatusPhase(cdiv1.CSICloneInProgress, datavolume, nil, CsiClone)
```

Comment:
This requeue necessary?

Reply:
I was under impression, that when we do Clone, we might want to requeue the same way as in Import to check progress/update status, but no I think that for CSI Clone it does not make any sense, there is nothing to reconcile until PVC is either Bound or Lost (or has some warning event)

Removed

Reply:
Looks like it does not work without requeue here, investigating.

Reply:
Some coordination between Namespace transfer controller and datavolume controller is probably not handled explicitely in this case, when I add Requeue it works, but I think I can be implement in a better way.

Reply:
finally got it, not needed, there was a wrong owner was set on transfer/temporary pvc (a DV from different namespace), and it was being deleted. I am not sure if k8s GC was deleting it or our logic in some other branch of code, but now after fixing that problem it works.